### ISSUE TITLE
模型问题

### ISSUE THREAD

**Issue by Lmonster** [LinXueyuanStdio/LaTeX_OCR#5]
我跑了一下full分支的数据，在第2个epoch之后就开始过拟合了。是否参数设置有问题呢？你是否跑完了全量数据呢？

**Comment by MTaoYue** [comment]
我也遇到了这个问题

**Comment by MTaoYue** [comment]
第二个epoch之后准确率都为0，训到了200个也是这样

**Comment by LinXueyuanStdio** [comment]
我也没跑完全部数据，因为我的机器太差，只有2G，只选了约2万来跑，效果就是REAMDME里的
训练时建议 batchsize 取 1，learning rate<0.1，2G机器跑2万，训练慢，但是准确率高
需要注意的是，在训练后期要保证 learning rate<0.001，不然后面loss会爆炸，跑多少 epoch 都没用（注意理解training.json，里面定义的learning rate的控制参数）

**Comment by MTaoYue** [comment]
感谢作者。以下是我做过的尝试中最好的结果：
 "export_name": "training.json",
    "device": "cuda:5",
    "criterion_method": "CrossEntropyLoss",
    "n_epochs": 400,
    "batch_size": 4,
    "dropout": 1,
    "clip": -1,
    "lr_method": "Adam",
    "lr_scheduler": "CosineAnnealingLR",
    "lr_init": 1e-4,
    "lr_min": 1e-6,
    "start_decay": 3,
    "end_decay": 20,
    "lr_warm": 1e-5,
    "end_warm": 2
2019-08-19 21:03:38,238:INFO: - Eval:  BLEU-4 is 87.62  ||  ExactMatchScore is 33.24  ||  EditDistance is 91.13  ||  perplexity is -1.17 
2019-08-19 21:03:38,238:INFO: - Elapsed time: 3733.46, learning rate: 0.00001
准确率最高为33.24，再往下训，loss下降，准确率也下降，过拟合

**Comment by Lmonster** [comment]
> 感谢作者。以下是我做过的尝试中最好的结果：
> "export_name": "training.json",
> "device": "cuda:5",
> "criterion_method": "CrossEntropyLoss",
> "n_epochs": 400,
> "batch_size": 4,
> "dropout": 1,
> "clip": -1,
> "lr_method": "Adam",
> "lr_scheduler": "CosineAnnealingLR",
> "lr_init": 1e-4,
> "lr_min": 1e-6,
> "start_decay": 3,
> "end_decay": 20,
> "lr_warm": 1e-5,
> "end_warm": 2
> 2019-08-19 21:03:38,238:INFO: - Eval: BLEU-4 is 87.62 || ExactMatchScore is 33.24 || EditDistance is 91.13 || perplexity is -1.17
> 2019-08-19 21:03:38,238:INFO: - Elapsed time: 3733.46, learning rate: 0.00001
> 准确率最高为33.24，再往下训，loss下降，准确率也下降，过拟合

你好，你有尝试predict非test集合的截图吗？效果是否好呢？我这里模型出来的泛化较差。

**Comment by MTaoYue** [comment]
> > 感谢作者。以下是我做过的尝试中最好的结果：
> > "export_name": "training.json",
> > "device": "cuda:5",
> > "criterion_method": "CrossEntropyLoss",
> > "n_epochs": 400,
> > "batch_size": 4,
> > "dropout": 1,
> > "clip": -1,
> > "lr_method": "Adam",
> > "lr_scheduler": "CosineAnnealingLR",
> > "lr_init": 1e-4,
> > "lr_min": 1e-6,
> > "start_decay": 3,
> > "end_decay": 20,
> > "lr_warm": 1e-5,
> > "end_warm": 2
> > 2019-08-19 21:03:38,238:INFO: - Eval: BLEU-4 is 87.62 || ExactMatchScore is 33.24 || EditDistance is 91.13 || perplexity is -1.17
> > 2019-08-19 21:03:38,238:INFO: - Elapsed time: 3733.46, learning rate: 0.00001
> > 准确率最高为33.24，再往下训，loss下降，准确率也下降，过拟合
> 
> 你好，你有尝试predict非test集合的截图吗？效果是否好呢？我这里模型出来的泛化较差。

试过，全是[[[[[[]]]]]]。

**Comment by ZhigangPu** [comment]
> > 感谢作者。以下是我做过的尝试中最好的结果：
> > "export_name": "training.json",
> > "device": "cuda:5",
> > "criterion_method": "CrossEntropyLoss",
> > "n_epochs": 400,
> > "batch_size": 4,
> > "dropout": 1,
> > "clip": -1,
> > "lr_method": "Adam",
> > "lr_scheduler": "CosineAnnealingLR",
> > "lr_init": 1e-4,
> > "lr_min": 1e-6,
> > "start_decay": 3,
> > "end_decay": 20,
> > "lr_warm": 1e-5,
> > "end_warm": 2
> > 2019-08-19 21:03:38,238:INFO: - Eval: BLEU-4 is 87.62 || ExactMatchScore is 33.24 || EditDistance is 91.13 || perplexity is -1.17
> > 2019-08-19 21:03:38,238:INFO: - Elapsed time: 3733.46, learning rate: 0.00001
> > 准确率最高为33.24，再往下训，loss下降，准确率也下降，过拟合
> 
> 你好，你有尝试predict非test集合的截图吗？效果是否好呢？我这里模型出来的泛化较差。

非常差+1
\begin{array} { c c } { { { { } & { } & { } & { } & { } & { } & { } & { } & { } & { } & { } & { } & { } & { } & { } & { } & { } & { } & { } & { } & { } & { } & { } & { } & { } & { } & { } & { } & { } & { } & { } & { } & { } & { } & { } & { } & { } & { } & { } & { } & { } & { } & { } & { } & { } & { } & { } & { } &

**Comment by ZhigangPu** [comment]
> 感谢作者。以下是我做过的尝试中最好的结果：
> "export_name": "training.json",
> "device": "cuda:5",
> "criterion_method": "CrossEntropyLoss",
> "n_epochs": 400,
> "batch_size": 4,
> "dropout": 1,
> "clip": -1,
> "lr_method": "Adam",
> "lr_scheduler": "CosineAnnealingLR",
> "lr_init": 1e-4,
> "lr_min": 1e-6,
> "start_decay": 3,
> "end_decay": 20,
> "lr_warm": 1e-5,
> "end_warm": 2
> 2019-08-19 21:03:38,238:INFO: - Eval: BLEU-4 is 87.62 || ExactMatchScore is 33.24 || EditDistance is 91.13 || perplexity is -1.17
> 2019-08-19 21:03:38,238:INFO: - Elapsed time: 3733.46, learning rate: 0.00001
> 准确率最高为33.24，再往下训，loss下降，准确率也下降，过拟合

看了Harvard NLP的论文，其结果EM可以达到87。

**Comment by LinXueyuanStdio** [comment]
首先，batchSize一般取1，因为看图写话的模型几乎都是1，算是炼丹经验吧

其次，泛化性能的话，还是要跑test，训练时的EM不能说明问题。并且重复符号多的预测明显是欠拟合，可能batchSize取4还是大了，数据量不能满足模型。

最后，Harvard NLP的论文是真的强，我最多EM到51(test)，明明结构差不多:sob:。这可能和本项目用tf实现、输入图片支持到1600x1600有关。本项目模型本身并没有写死padding到一个确定的数，而是给了一系列bucket，向上padding（比如buncket=[100x200，200x200]，输入img=90x200 padding到100x200，img=110x200 padding到200x200），在大img的输入下EM没有明显下降。我也尝试过在pytorch中使用bucket，但是失败了，pytorch就是要写死输入张量大小。在pytorch里padding到一个大的数，训练超级慢。或许pytorch有实现bucket的方法但是我没找到，如果您知道如何实现的，请务必留个demo项目地址，大家一起观摩。或者有完美复现Harvard论文的兄弟，也可以分享一下心得。

**Comment by Junwei9103** [comment]
我想问下模型预测输入的问题：
问题描述：
LaTex代码：\sqrt{a^{2}+b^{2}} + kx_{j} + \alpha_{i}
运行python predict.py进行预测
1 首先我用matplotlib画出了上面的LaTex代码，显示出公式，然后截图公式部分，把截图（.png）拿去做预测，输出的预测结果[[[[[[]]]]]]差不多都是这些无效的东西。
2 换一种方案，把上面的LaTex代码放到small.formulas/test.norm.txt文件中，python build.py生成test图片，然后再把生成的图片拿去predict.py预测，结果为：\sqrt{a^{2} + b^{2} + k_{j} + \alpha_{j}}(虽然结果不是很准确，但是可以看出是预期的结果，毕竟只跑了一个epoch)。

是不是模型输入的图片有什么要求呢？固定尺寸吗？

**Comment by LinXueyuanStdio** [comment]
@Junwei9103 

事实上模型的输入还需要预处理。
predict.py里没有包含预处理的部分。它对一般的图片直接读取然后直接丢给模型，以致内容平移对模型造成了影响。
build.py自然地包含了预处理的部分，所以你两种方式的输出一个异常一个正常。

要修复这个问题，可以参考model.utils.images下相关代码。预处理我在这里大概描述一下。
1. crop_image 把空白部分去掉，裁剪剩公式内容
2. pad_image 加空白填充，上下左右填充 8 个像素后用 bucket 确定最终大小，左上角空 8 个像素，填入公式内容的那些像素
3. downsample_image 下采样

**Comment by Junwei9103** [comment]
@LinXueyuanStdio 好的，我再去试一下，添加预处理。非常感谢！

**Comment by PKU-FengShuo** [comment]
在predict.py中添加了预处理，添加了二值化，还添加了crop_image，pad_image，downsample_image，build.py生成的图片
![0](https://user-images.githubusercontent.com/20044866/74611485-428fe880-5137-11ea-8685-6a6bfd2a8f64.png)
识别为
W = \{ 2 , 3 , 4 , 8 , 8 , 9 \}
预处理后的图片
![test_new_](https://user-images.githubusercontent.com/20044866/74611504-8a167480-5137-11ea-8cc8-1f2e3f1d13ef.png)
识别为
{ { W } = ^ { , } ( 2 , \ \ \chi _ { + } , \ \gamma _ { 1 } , \ 5 , \ 9 )
我训练了40epoch，请问我还落下什么了吗

**Comment by ccl-private** [comment]
@PKU-FengShuo 应该是对数据过拟合了，我训练了一个模型，没有添加了crop_image，pad_image，downsample_image也可以识别部分公式。可见模型对图片尺寸还是有一定宽容度的。所以还是考虑模型过拟合的成分比较大，我现在正在测试开启droupout减少过拟合，我先训练看看效果。
